### ISSUE TITLE
i18n -with subdirectories, like www.site.com/en

### ISSUE THREAD

**Issue by heyt0ny** [kataras/iris#1369]
Hello!

I'm using `i18n middleware` for lang localization and currently my urls look like `www.site.com/?lang=en-US`. According to many sources (e.g. https://webmasters.stackexchange.com/questions/403/how-should-i-structure-my-urls-for-both-seo-and-localization) it's much better to use subdirectories instead of query params. So how to make 

 `www.site.com/?lang=en-US` ->  `www.site.com/en-US/`

Thanks!

**Comment by kataras** [comment]
Hello @heyt0ny,

What i18n middleware are you using?

- https://github.com/kataras/iris/tree/master/middleware/i18n
- https://github.com/iris-contrib/middleware/tree/master/go-i18n

With both of them you can **specify the way that a language is accepted**.
See for example the kataras/iris/middleware/i18n will try to get the language first by language's context's key(can be customized too). 

https://github.com/kataras/iris/blob/df882273e21952a316236174123fc09096b49aad/configuration.go#L717-L718

https://github.com/kataras/iris/blob/df882273e21952a316236174123fc09096b49aad/middleware/i18n/i18n.go#L22-L23



```go
setLangMiddleware := func(ctx iris.Context){
    langKey := ctx.Application().ConfigurationReadOnly().GetTranslateLanguageContextKey()
    languageByPath := ctx.Params().Get("lang") // see {lang}
    ctx.Values().Set(langKey, languageByPath)
    ctx.Next()
}
// app or per group (routes := app.Party; routes.Use...)
app.Use(setLangMiddleware)
app.Use(theI18nMiddlewareInstance)
app.Get("/{lang}/path", yourHandler)
```

**Comment by heyt0ny** [comment]
Thank you, and if I want to apply {lang} to many parties (P) and routes (R)- I should create new party with `{lang}` route and put P & R into it ?

**Comment by kataras** [comment]
@heyt0ny You are welcome, how do you that with net/http package or other framework? It depends what fits you the most, I can help you convert everything to Iris.

**Comment by kataras** [comment]
@heyt0ny You are welcome.You can do it understand by mapping the locale keys with the param and if it is NOT matching then redirect to the default (en-us or without param) otherwise handle the request with the specific locale. You can make use of `ctx.Values()` inside the app's middleware and call the necessary handlers.

However, you can describe us the net/http (or other framework at any language) way because it depends on what fits you the most, I can help you convert everything to Iris.

If you don't have any idea about it, give me time until tomorrow and I will provide such a feature (if necessary).

**Comment by kataras** [comment]
Hello @heyt0ny,

If I may understand, you need to
- register the routes **once for all languages**
- if the path is prefixed with one of the languages codes, e.g. (`zh-cn` and  `el-gr`  or  `en-us`) such as `http://example.com/zh-cn/some-path` then "redirect to / execute handler(s) of "(**local redirect to**) `/some-path` with the `lang` set to `zh-CN`.

A new [router wrapper](https://github.com/kataras/iris/blob/655f649a831c56177652fc0ddbd5870b64970a05/_examples/routing/custom-wrapper/main.go#L36) for that task **will be included on the next minor version of Iris 12.0.2**.

![](https://media.giphy.com/media/Ur7hcyvhet3z6jxKel/giphy.gif)

Please check it out and share your thoughts with us:

```go
import (
    "net/http"

    "github.com/kataras/iris/v12/middleware/i18n"
)

// [...]

func NewWrapper(c i18n.Config) func(http.ResponseWriter, *http.Request, http.HandlerFunc) {
	return func(w http.ResponseWriter, r *http.Request, routerHandler http.HandlerFunc) {
		path := strings.ToLower(r.URL.Path[1:])

		if idx := strings.IndexRune(path, '/'); idx > 0 {
			path = path[:idx]
		}

		for lang := range c.Languages {
			if strings.ToLower(lang) == path {
				path = r.URL.Path[len(path)+1:]
				if path == "" {
					path = "/"
				}
				r.RequestURI = path
				r.URL.Path = path
				r.Header.Set("Accept-Language", lang)
			}
		}

		routerHandler(w, r)
	}
}
```

And **usage**:

```go
import (
    "github.com/kataras/iris/v12"
    "github.com/kataras/iris/v12/middleware/i18n"
)

// [...]

i18nConfig := i18n.Config{
	Default:       "en-US",
	URLParameter:  "lang",
	Languages: map[string]string{
		"en-US": "./locales/locale_en-US.ini",
		"el-GR": "./locales/locale_el-GR.ini",
		"zh-CN": "./locales/locale_zh-CN.ini",
	},
}
app.WrapRouter(NewWrapper(i18nConfig)) // <- HERE

i18nMiddleware := i18n.New(i18nConfig) // <- HERE
app.Use(i18nMiddleware)
```

```go
app.Get("/", func(ctx iris.Context) {
	hi := ctx.Translate("hi", "iris")

	language := ctx.Values().GetString(
	    // GetTranslateLanguageContextKey() == "language"
            ctx.Application().ConfigurationReadOnly().GetTranslateLanguageContextKey()
        )

	ctx.Writef("From the language %s translated output: %s", language, hi)
})

app.Get("/some-path", func(ctx iris.Context) {
	ctx.Writef("%s", ctx.Translate("hi", "iris"))
})


```

**Comment by heyt0ny** [comment]
That's great! I think it's exactly what I was looking for. Thank you 🙏

**Comment by kataras** [comment]
OK, it's on v12 branch by [this commit](https://github.com/kataras/iris/commit/b61d07841a5aee63d231ca889f9ec6209dac64e0) which will be released at 23 November under v12.0.2.

The above commit contains some fixes and improvements such as `config.SetCookie`, `config.Alternatives`, find and match `"en"` as`"en-US"`, `ctx.TranslateLang("zh-CN", "word" args...)` which is a helper that translates a specific word by given locale(first parameter) - useful for things like language menu selection (we still have `ctx.Translate("word", args...)` for current locale translate that can be passed on templates too) and e.t.c.

I am very grateful for your contribution and this feature request, it is quite unique @heyt0ny .

**Comment by heyt0ny** [comment]
Sounds cool! You are developing really great framework. Keep it up 👍
